### ISSUE TITLE
snpEff-4.0e disabled our junit4.10 +, it make  TemporaryFolder rule work improperly. 

### ISSUE THREAD

**Issue by ChristinaXu2017** [AdamaJava/adamajava#92]
Under unit tests,  the testing output should be deleted once the test method finishes,  org.junit.rules.TemporaryFolder should execute this task well.  

the code looks like: 

```
import org.junit.Rule;
public class DbsnpModeTest {
@Rule
public final TemporaryFolder testFolder = new TemporaryFolder();
@Test
public void annotationTest() throws Exception{
File fvcf = testFolder.newFile();
```

the build file looks like:
```
testCompile 'junit:junit:4.10' //or
testCompile 'junit:junit:4.12'
```

Junit 4.10 works well but junit:4.12 throw below exception:
java.lang.IllegalStateException: the temporary folder has not yet been created

**Comment by ChristinaXu2017** [comment]
I have to work in many different ways, change JUnit from 4.4 to 4.12. debug and trace the code and error messages. 

I realize TemporaryFolder.before() is not called  In both junit:junit:4.12 and 4.10; the variable "folder" is "null" in both version. however method newFile() in junit:4.10 accept default null value. the code details are below: 

------- (junit:4.12) code: "folder == null"  and getRoot() throw exception ------
public File newFile(){ return File.createTempFile("junit", null, getRoot());}
public File getRoot() {
        if (folder == null) {   throw new IllegalStateException(  "the temporary folder has not yet been created");   }
        return folder;
    }
--------- (junit:4.12) code: before() is protected, it is not called ----------------
@Override
protected void before() throws Throwable {  create();   }
public void create() throws IOException {  folder = createTemporaryFolderIn(parentFolder);  }
private File createTemporaryFolderIn(File parentFolder) throws IOException {
        File createdFolder = File.createTempFile("junit", "", parentFolder);
        ....
        return createdFolder;
    }

But related code in  junit:junit:4.10 is different
--------(junit:4.10) code:  "folder == null" but File.createTempFile accept null value -----------
public File newFile() throws IOException { return File.createTempFile("junit", null, folder); }

--------- (junit:4.10) code: before() is protected, it is not called ----------------
@Override
protected void before() throws Throwable { create();  }
public void create() throws IOException {  folder= newFolder();  }
public File newFolder() throws IOException {
		File createdFolder= File.createTempFile("junit", "", folder);
		....
		return createdFolder;
	}

**Comment by ChristinaXu2017** [comment]
both JUnit version works well in eclipse. JUnit 4.12 complain under Gradle test but 4.10 is OK even the before() method is not called.

**Comment by ChristinaXu2017** [comment]
it looks like @Rule @Befor @After are ignored during Gradle test.  I may try to use BlockJUnit4ClassRunner according to below discuss: 
https://nuget.pkg.github.com/gradle/exemplar/issues/42

**Comment by ChristinaXu2017** [comment]
It is obvious that Junit new feature @rule @Befor @after are ignored during Gradle test. The default runner in Gradle is "JUnit4ClassRunner" but not "BlockJUnit4ClassRunner". See below Exception message from Gradle run:

> java.lang.IllegalStateException: the temporary folder has not yet been created
> 	at org.junit.rules.TemporaryFolder.getRoot(TemporaryFolder.java:145)
>       ...
>      at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:42)
> 	at org.junit.internal.runners.JUnit4ClassRunner.invokeTestMethod(JUnit4ClassRunner.java:88)
> 	at org.junit.internal.runners.JUnit4ClassRunner.runMethods(JUnit4ClassRunner.java:51)
>       ...

To make TemporaryFolder rule work, we have to call BlockJUnit4ClassRunner.

**Comment by holmeso** [comment]
I may well be missing something here, but why don't we just stick to 4.10?

**Comment by ChristinaXu2017** [comment]
4.12 is better than 4.10, but Gradle dislike 4.12, we have to temporary reverse back to 4.10 to avoid the exception.

**Comment by ChristinaXu2017** [comment]
I have created a runner, then pass to each test class @RunWith(CallBlockJUnit4ClassRunner.class ).  Then Gradle should point to this runner as default one, then @rule @Befor @after will be picked up. The previous exception will vanish due to  "TemporaryFolder::before()" is called,  the "private File folder;" will be initialized.

**Comment by holmeso** [comment]
> 4.12 is better than 4.10, but Gradle dislike 4.12, we have to temporary reverse back to 4.10 to avoid the exception.

Can you provide details as to why 4.12 is better?
Do we have tests that only work with 4.12?
I think to make a move like this, there needs to be a compelling reason other than that it is better.
I would be inclined to leave our code working with 4.10, and once we up our gradle version, we move to junit 5

**Comment by ChristinaXu2017** [comment]
Whether 4.12 is better than 4.10, it is not up to my judgment, there is a reason why they update from 4.10 to 4.12. 
My personal experience is that the Gradle (I tried several different version) can't pick up JUnit @Rule annotation. 4.10 keep it silent but 4.12 report an exception. 

The purpose of @Rule is allowing us to do something before a test method is run and after a test method has been run.  But it must run with CallBlockJUnit4ClassRunner.  In Gradle the default runner is JUnit4ClassRunner even in newer gradle version, it didn't do something before/after a test method. 

If you try 
```
@Rule
public  TemporaryFolder testFolder = new TemporaryFolder();
...
testFolder.newFile();   //(A)
testFolder.newFile("anewfile.name");  // (B)
```
Both 4.10 and 4.12 will throw an exception when they arrive "(B)", because the "private File folder;" inside calss TemporaryFolder is not initalized due to @before() method is not called.  The reason 4.10 didn't throw exception on "(A)"  is that it created another TempFile() but not under "folder".  4.12 tried to stop this option, it force "(A)" have to create under "folder".

**Comment by ChristinaXu2017** [comment]
I am not sure whether Gradle will work with junit 5 better, It is very possible more new problem occurs during new versions.  

rule of TemporaryFolder is very common and widely used.  I already created a customer runner which tells Gradle to pick up CallBlockJUnit4ClassRunner and make "@Rule" work well in our Gradle.  It is in my pull request.

**Comment by ChristinaXu2017** [comment]
This issue is solved by pull request #97, the reason is as described in the title. See the details on PR #97
